packages/plugin-node-modules/sources/NodeModulesLinker.ts

```diff
@@ -408,10 +408,43 @@ const copyPromise = async (dstDir: PortablePath, srcDir: PortablePath, {baseFs}:
   }
 };
 
+/**
+ * This function removes node_modules roots that do not exist on the filesystem from the location tree.
+ *
+ * This is needed to transparently support workflows on CI systems. When
+ * user caches only top-level node_modules and forgets to cache node_modules
+ * from deeper workspaces. By removing non-existent node_modules roots
+ * we make our location tree to represent the real tree on the file system.
+ *
+ * Please note, that this function doesn't help with any other inconsistency
+ * on a deeper level inside node_modules tree, it helps only when some node_modules roots
+ * do not exist at all
+ *
+ * @param locationTree location tree
+ *
+ * @returns location tree with non-existent node_modules roots stripped
+ */
+async function refineLocationRoots(locationTree: LocationTree): Promise<LocationTree> {
+  const refinedTree: LocationTree = new Map();
+
+  const promises = [];
+  for (const [nodeModulesRoot, node] of locationTree.entries()) {
+    promises.push(async () => {
+      if (await xfs.existsPromise(nodeModulesRoot)) {
```

Comment:
Fwiw better use `existsSync` - small checks like this are typically faster when synchronous, and `exists` is semantically more sound when synchronous (because the window for the information to become stale is smaller).

Still good to merge since we have a bunch of `existsPromise` already - I'll get rid of them all in one pass.

Reply:
Yep, I agree, better to use sync, as the number of checks will be very few here

Reply:
Done